soc/xtensa/intel_adsp/tools/cavstool.py

```diff
@@ -659,30 +660,65 @@ async def _main(server):
         if dsp.HIPCIDA & 0x80000000:
             dsp.HIPCIDA = 1<<31 # must ACK any DONE interrupts that arrive!
 
+        if server:
+            # Check if the client connection is alive.
+            if not is_connection_alive(server):
+                lock.acquire()
+                start_output = False
+                lock.release()
+
 class adsp_request_handler(socketserver.BaseRequestHandler):
     """
     The request handler class for control the actions of server.
     """
 
-    def receive_fw(self, filename):
+    def receive_fw(self):
+        header_sz = 80
+
+        log.info("Receiving...")
+        # The first contain the header and the front
+        # data of the firmware
+        d = self.request.recv(BUF_SIZE)
+
+        # Unpacked the header data
+        # Include size(4), filename(42), MD5(32) and pad(2)
+        header = d[:header_sz]
+        total = d[header_sz:]
+        cnt = len(total)
+        s = struct.Struct('I 42s 32s H')
+        fsize, fname, md5_tx_b, pad = s.unpack(header)
+        log.info(f'size:{fsize}, filename:{fname}, MD5:{md5_tx_b}, pad:{pad})')
+
+
+        # Receive the remain part of the firmware. We only
+        # receive the specified amount of bytes.
+        while cnt < fsize:
+            data = self.request.recv(BUF_SIZE)
+            if not data:
```

Comment:
Please let me explain it: In the server-side, the recv() will be blocked until the client's socket.shutdown/close called. We hope the client can keep receiving the download result in the current connection (after completing MD5 check) from the server and return it to the west runner. So we only receive the specified bytes then keep running and inform the client.

For using python Remote Objects or JSONRPC and I think it's a good suggestion but I might need a little more time to realize it. Thank you.

Reply:
`not data` means the socket has been closed. Why would the socket be closed when there is still data to read? Only when something went wrong => exception.

Please describe what happens when you remove this `break`.

Reply:
hi @marc-hb , it seem it still work well if I remove the break.

Reply:
I think the break here can be remove.

Reply:
When I asked to remove the `break` it was only temporarily for testing purposes, not for the final code. You still need to check for disconnections otherwise you can have an infinite loop.